syntax = "proto3";

message Key {
  string key_name = 1;
}

message Value {
  string value_name = 1;
}

message KeyValue {
  Key key = 1;
  Value value = 2;
}

message Status {
  string result = 1;
}

message Deque {
  Key key_name = 1;
  repeated string data = 2;
}

message ListRange {
  Key key = 1;
  int32 start = 2;
  int32 stop = 3;
}

message ResponseVal {
  oneof res {
   bool noval = 1;
   Value value = 2;
  }
}

service KeyStore {
  rpc Set(KeyValue) returns (Status);
  rpc Get(Key) returns (ResponseVal);
  rpc Delete(Key) returns (Status);
  rpc LPush(KeyValue) returns (Status);
  rpc LRange(ListRange) returns (Deque);
  rpc LPop(Key) returns (ResponseVal);
  // rpc ListAll(Key) returns (stream KeyValue);
}

